File: rust/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::io::{self, Read, Write};

pub const BINARY_ASSET_RESPONSE_MAGIC: &[u8; 4] = b"FBRB";
pub const BINARY_BUILD_RESULT_MAGIC: &[u8; 4] = b"FBRR";
pub const MAX_FRAME_LENGTH: usize = 256 * 1024 * 1024;

#[derive(Debug, Serialize, Deserialize)]
pub struct BuildOutput {
    pub asset: String,
    pub bytes: Vec<u8>,
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Diagnostic {
    pub level: String,
    pub message: String,
    #[serde(default)]
    pub asset: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd)]
pub struct GlobRead {
    pub package: String,
    pub pattern: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BuildResult {
    #[serde(rename = "type")]
    pub message_type: String,
    pub id: u64,
    pub status: String,
    #[serde(default)]
    pub outputs: Vec<BuildOutput>,
    #[serde(default)]
    pub deleted: Vec<String>,
    #[serde(default)]
    pub reads: Vec<String>,
    #[serde(default)]
    pub resolver_reads: Vec<String>,
    #[serde(default)]
    pub glob_reads: Vec<GlobRead>,
    #[serde(default)]
    pub diagnostics: Vec<Diagnostic>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug)]
pub struct BinaryFrame {
    pub magic: [u8; 4],
    pub metadata: Value,
    payload: Vec<u8>,
    bytes_start: usize,
}

#[derive(Debug)]
pub enum IncomingFrame {
    Json(Value),
    Binary(BinaryFrame),
}

pub fn write_frame<W: Write, T: Serialize>(writer: &mut W, message: &T) -> io::Result<usize> {
    let payload = serde_json::to_vec(message).map_err(io::Error::other)?;
    write_payload(writer, &payload)
}

pub fn write_binary_frame<W: Write, T: Serialize>(
    writer: &mut W,
    metadata: &T,
    bytes: &[u8],
) -> io::Result<usize> {
    write_binary_frame_with_magic(writer, BINARY_ASSET_RESPONSE_MAGIC, metadata, bytes)
}

pub fn write_binary_frame_with_magic<W: Write, T: Serialize>(
    writer: &mut W,
    magic: &[u8; 4],
    metadata: &T,
    bytes: &[u8],
) -> io::Result<usize> {
    let metadata = serde_json::to_vec(metadata).map_err(io::Error::other)?;
    let payload_length = magic
        .len()
        .checked_add(4)
        .and_then(|length| length.checked_add(metadata.len()))
        .and_then(|length| length.checked_add(bytes.len()))
        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "IPC frame is too large"))?;
    if payload_length > MAX_FRAME_LENGTH {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "IPC frame exceeds the 256 MiB safety limit",
        ));
    }
    let metadata_length = u32::try_from(metadata.len()).map_err(|_| {
        io::Error::new(io::ErrorKind::InvalidData, "IPC metadata is too large")
    })?;
    let length = u32::try_from(payload_length)
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "IPC frame is too large"))?;
    writer.write_all(&length.to_be_bytes())?;
    writer.write_all(magic)?;
    writer.write_all(&metadata_length.to_be_bytes())?;
    writer.write_all(&metadata)?;
    writer.write_all(bytes)?;
    writer.flush()?;
    Ok(payload_length + 4)
}

fn write_payload<W: Write>(writer: &mut W, payload: &[u8]) -> io::Result<usize> {
    if payload.len() > MAX_FRAME_LENGTH {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "IPC frame exceeds the 256 MiB safety limit",
        ));
    }
    let length = u32::try_from(payload.len())
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "IPC frame is too large"))?;
    writer.write_all(&length.to_be_bytes())?;
    writer.write_all(payload)?;
    writer.flush()?;
    Ok(payload.len() + 4)
}

pub fn read_message_with_size<R: Read>(
    reader: &mut R,
) -> io::Result<Option<(IncomingFrame, usize)>> {
    let mut header = [0_u8; 4];
    match reader.read_exact(&mut header) {
        Ok(()) => {}
        Err(error) if error.kind() == io::ErrorKind::UnexpectedEof => return Ok(None),
        Err(error) => return Err(error),
    }

    let length = u32::from_be_bytes(header) as usize;
    if length > MAX_FRAME_LENGTH {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "IPC frame exceeds the 256 MiB safety limit",
        ));
    }

    let mut payload = vec![0_u8; length];
    reader.read_exact(&mut payload)?;
    Ok(Some((decode_payload(payload)?, length + 4)))
}

fn decode_payload(payload: Vec<u8>) -> io::Result<IncomingFrame> {
    if payload.starts_with(BINARY_ASSET_RESPONSE_MAGIC)
        || payload.starts_with(BINARY_BUILD_RESULT_MAGIC)
    {
        if payload.len() < 8 {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "binary IPC envelope is truncated",
            ));
        }
        let mut magic = [0_u8; 4];
        magic.copy_from_slice(&payload[..4]);
        let metadata_length = u32::from_be_bytes(payload[4..8].try_into().unwrap()) as usize;
        let metadata_start = 8_usize;
        let bytes_start = metadata_start
            .checked_add(metadata_length)
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "IPC metadata is too large"))?;
        if bytes_start > payload.len() {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "binary IPC metadata is truncated",
            ));
        }
        let metadata = serde_json::from_slice(&payload[metadata_start..bytes_start])
            .map_err(io::Error::other)?;
        return Ok(IncomingFrame::Binary(BinaryFrame {
            magic,
            metadata,
            payload,
            bytes_start,
        }));
    }

    serde_json::from_slice(&payload)
        .map(IncomingFrame::Json)
        .map_err(io::Error::other)
}

#[derive(Debug, Deserialize)]
struct BinaryBuildOutput {
    asset: String,
    length: u64,
}

#[derive(Debug, Deserialize)]
struct BinaryBuildResultMetadata {
    #[serde(rename = "type")]
    message_type: String,
    id: u64,
    status: String,
    #[serde(default)]
    encoding: Option<String>,
    #[serde(default)]
    outputs: Vec<BinaryBuildOutput>,
    #[serde(default)]
    deleted: Vec<String>,
    #[serde(default)]
    reads: Vec<String>,
    #[serde(default)]
    resolver_reads: Vec<String>,
    #[serde(default)]
    glob_reads: Vec<GlobRead>,
    #[serde(default)]
    diagnostics: Vec<Diagnostic>,
    #[serde(default)]
    error: Option<String>,
}

impl BinaryBuildResultMetadata {
    fn decode(self, bytes: &[u8], cursor: &mut usize) -> io::Result<BuildResult> {
        if self.message_type != "build_result" {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "binary frame is not a build result",
            ));
        }
        if self.encoding.as_deref() != Some("raw") {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "build result does not use raw output encoding",
            ));
        }

        let mut outputs = Vec::with_capacity(self.outputs.len());
        for output in self.outputs {
            let length = usize::try_from(output.length).map_err(|_| {
                io::Error::new(io::ErrorKind::InvalidData, "build output is too large")
            })?;
            let end = cursor
                .checked_add(length)
                .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "build output overflows"))?;
            if end > bytes.len() {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    "build output bytes are truncated",
                ));
            }
            outputs.push(BuildOutput {
                asset: output.asset,
                bytes: bytes[*cursor..end].to_vec(),
            });
            *cursor = end;
        }

        Ok(BuildResult {
            message_type: self.message_type,
            id: self.id,
            status: self.status,
            outputs,
            deleted: self.deleted,
            reads: self.reads,
            resolver_reads: self.resolver_reads,
            glob_reads: self.glob_reads,
            diagnostics: self.diagnostics,
            error: self.error,
        })
    }
}

pub fn decode_build_result_frame(frame: BinaryFrame) -> io::Result<BuildResult> {
    if frame.magic != *BINARY_BUILD_RESULT_MAGIC {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "unexpected binary frame for build result",
        ));
    }
    let BinaryFrame {
        metadata,
        payload,
        bytes_start,
        ..
    } = frame;
    let metadata: BinaryBuildResultMetadata =
        serde_json::from_value(metadata).map_err(io::Error::other)?;
    let bytes = &payload[bytes_start..];
    let mut cursor = 0;
    let result = metadata.decode(bytes, &mut cursor)?;
    if cursor != bytes.len() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "build result contains trailing output bytes",
        ));
    }
    Ok(result)
}

#[derive(Debug, Deserialize)]
struct BinaryBuildBatchResultMetadata {
    #[serde(rename = "type")]
    message_type: String,
    id: u64,
    #[serde(default)]
    encoding: Option<String>,
    results: Vec<BinaryBuildResultMetadata>,
}

pub struct DecodedBuildBatchResult {
    pub id: u64,
    pub results: Vec<BuildResult>,
}

pub fn decode_build_batch_result_frame(frame: BinaryFrame) -> io::Result<DecodedBuildBatchResult> {
    if frame.magic != *BINARY_BUILD_RESULT_MAGIC {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "unexpected binary frame for build batch result",
        ));
    }
    let BinaryFrame {
        metadata,
        payload,
        bytes_start,
        ..
    } = frame;
    let metadata: BinaryBuildBatchResultMetadata =
        serde_json::from_value(metadata).map_err(io::Error::other)?;
    if metadata.message_type != "build_batch_result" {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "binary frame is not a build batch result",
        ));
    }
    if metadata.encoding.as_deref() != Some("raw") {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "build batch result does not use raw output encoding",
        ));
    }

    let bytes = &payload[bytes_start..];
    let mut cursor = 0;
    let results = metadata
        .results
        .into_iter()
        .map(|result| result.decode(bytes, &mut cursor))
        .collect::<io::Result<Vec<_>>>()?;
    if cursor != bytes.len() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "build batch result contains trailing output bytes",
        ));
    }
    Ok(DecodedBuildBatchResult {
        id: metadata.id,
        results,
    })
}

#[cfg(test)]
mod tests {
    use super::{
        decode_build_batch_result_frame, decode_build_result_frame, read_message_with_size,
        write_binary_frame, write_binary_frame_with_magic, write_frame, IncomingFrame,
        BINARY_ASSET_RESPONSE_MAGIC, BINARY_BUILD_RESULT_MAGIC, MAX_FRAME_LENGTH,
    };
    use serde_json::{json, Value};
    use std::io::Cursor;

    #[test]
    fn frame_round_trip_uses_length_prefix() {
        let message = json!({"v": 1, "type": "test", "unicode": "日本語"});
        let mut bytes = Vec::new();
        write_frame(&mut bytes, &message).expect("write frame");

        let payload_length = u32::from_be_bytes(bytes[..4].try_into().unwrap());
        assert_eq!(payload_length as usize, bytes.len() - 4);
        let (frame, frame_size) = read_message_with_size(&mut Cursor::new(bytes))
            .unwrap()
            .expect("frame");
        let IncomingFrame::Json(decoded) = frame else {
            panic!("expected JSON frame");
        };
        assert_eq!(decoded, message);
        assert_eq!(frame_size, payload_length as usize + 4);
    }

    #[test]
    fn binary_frame_contains_metadata_and_raw_bytes_in_one_payload() {
        let metadata = json!({
            "v": 1,
            "type": "asset_response",
            "id": 1000,
            "ok": true,
            "encoding": "raw",
            "length": 7,
        });
        let raw_bytes = b"payload";
        let mut bytes = Vec::new();
        write_binary_frame(&mut bytes, &metadata, raw_bytes).expect("write binary frame");

        let payload_length = u32::from_be_bytes(bytes[..4].try_into().unwrap()) as usize;
        assert_eq!(payload_length, bytes.len() - 4);
        assert_eq!(&bytes[4..8], BINARY_ASSET_RESPONSE_MAGIC);
        let metadata_length = u32::from_be_bytes(bytes[8..12].try_into().unwrap()) as usize;
        let metadata_start = 12;
        let raw_start = metadata_start + metadata_length;
        assert_eq!(
            serde_json::from_slice::<Value>(&bytes[metadata_start..raw_start]).unwrap(),
            metadata
        );
        assert_eq!(&bytes[raw_start..], raw_bytes);
    }

    #[test]
    fn build_result_binary_frame_preserves_raw_outputs() {
        let metadata = json!({
            "v": 1,
            "type": "build_result",
            "id": 2,
            "status": "success",
            "encoding": "raw",
            "outputs": [{"asset": "app|lib/model.g.dart", "length": 7}],
            "reads": [],
            "resolver_reads": [],
            "glob_reads": [],
            "diagnostics": []
        });
        let raw_bytes = b"payload";
        let mut bytes = Vec::new();
        write_binary_frame_with_magic(
            &mut bytes,
            BINARY_BUILD_RESULT_MAGIC,
            &metadata,
            raw_bytes,
        )
        .expect("write build result binary frame");

        let (frame, frame_size) = read_message_with_size(&mut Cursor::new(bytes))
            .unwrap()
            .expect("frame");
        let IncomingFrame::Binary(frame) = frame else {
            panic!("expected binary frame");
        };
        assert_eq!(frame.magic, *BINARY_BUILD_RESULT_MAGIC);
        assert_eq!(&frame.payload[frame.bytes_start..], raw_bytes);
        assert_eq!(frame_size, 4 + 4 + 4 + frame.metadata.to_string().len() + raw_bytes.len());

        let result = decode_build_result_frame(frame).expect("decode build result");
        assert_eq!(result.id, 2);
        assert_eq!(result.outputs[0].asset, "app|lib/model.g.dart");
        assert_eq!(result.outputs[0].bytes, raw_bytes);
    }

    #[test]
    fn build_batch_binary_frame_restores_all_output_ranges() {
        let metadata = json!({
            "v": 1,
            "type": "build_batch_result",
            "id": 7,
            "encoding": "raw",
            "results": [
                {
                    "v": 1,
                    "type": "build_result",
                    "id": 0,
                    "status": "success",
                    "encoding": "raw",
                    "outputs": [{"asset": "app|a.g.dart", "length": 3}],
                    "reads": [],
                    "resolver_reads": [],
                    "glob_reads": [],
                    "diagnostics": []
                },
                {
                    "v": 1,
                    "type": "build_result",
                    "id": 1,
                    "status": "success",
                    "encoding": "raw",
                    "outputs": [{"asset": "app|b.g.dart", "length": 4}],
                    "reads": [],
                    "resolver_reads": [],
                    "glob_reads": [],
                    "diagnostics": []
                }
            ]
        });
        let raw_bytes = b"onefour";
        let mut bytes = Vec::new();
        write_binary_frame_with_magic(
            &mut bytes,
            BINARY_BUILD_RESULT_MAGIC,
            &metadata,
            raw_bytes,
        )
        .expect("write build batch result binary frame");

        let (frame, _) = read_message_with_size(&mut Cursor::new(bytes))
            .unwrap()
            .expect("frame");
        let IncomingFrame::Binary(frame) = frame else {
            panic!("expected binary frame");
        };
        let decoded = decode_build_batch_result_frame(frame).expect("decode build batch result");
        assert_eq!(decoded.id, 7);
        assert_eq!(decoded.results.len(), 2);
        assert_eq!(decoded.results[0].outputs[0].bytes, b"one");
        assert_eq!(decoded.results[1].outputs[0].bytes, b"four");
    }

    #[test]
    fn frame_length_limit_is_checked_before_allocation() {
        let length = (MAX_FRAME_LENGTH as u32 + 1).to_be_bytes();
        let error = read_message_with_size(&mut Cursor::new(length))
            .expect_err("oversized frame must be rejected");
        assert_eq!(error.kind(), std::io::ErrorKind::InvalidData);
    }
}

File: lib/src/protocol.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

typedef JsonMap = Map<String, dynamic>;
typedef RpcControlMessageHandler =
    Future<void> Function(WorkerBuildRequest message);

const List<int> _binaryAssetResponseMagic = <int>[0x42, 0x52, 0x41, 0x42];
const List<int> _binaryBuildResultMagic = <int>[0x42, 0x52, 0x41, 0x52];
const int _maxFrameLength = 256 * 1024 * 1024;

/// Typed messages received by the resident worker from the Rust frontend.
///
/// JSON is decoded and validated once at the IPC boundary. The execution
/// layer should consume these values instead of repeatedly inspecting maps.
sealed class WorkerMessage {
  const WorkerMessage({required this.id});

  final Object? id;

  static WorkerMessage decode(JsonMap message) {
    switch (message['type']) {
      case 'initialize':
        return WorkerInitializeMessage.fromJson(message);
      case 'reset':
        return WorkerResetMessage.fromJson(message);
      case 'reset_resolver':
        return WorkerResetResolverMessage.fromJson(message);
      case 'build':
        return WorkerBuildMessage(WorkerBuildRequest.fromJson(message));
      case 'build_batch':
        return WorkerBuildBatchMessage.fromJson(message);
      default:
        return UnsupportedWorkerMessage(
          id: message['id'],
          type: message['type'],
        );
    }
  }
}

class WorkerInitializeMessage extends WorkerMessage {
  WorkerInitializeMessage({
    required int id,
    required this.package,
    required this.phaseCount,
  }) : super(id: id);

  factory WorkerInitializeMessage.fromJson(JsonMap message) {
    final rawPhaseCount = message['phase_count'];
    final phaseCount = rawPhaseCount is num
        ? (rawPhaseCount.toInt() < 1 ? 1 : rawPhaseCount.toInt())
        : 1;
    final package = _requiredString(message, 'package', 'initialize');
    if (package.isEmpty) {
      throw const FormatException('initialize requires a non-empty package');
    }
    return WorkerInitializeMessage(
      id: _requiredInt(message, 'id', 'initialize'),
      package: package,
      phaseCount: phaseCount,
    );
  }

  final String package;
  final int phaseCount;
}

class WorkerResetMessage extends WorkerMessage {
  WorkerResetMessage({required int id}) : super(id: id);

  factory WorkerResetMessage.fromJson(JsonMap message) =>
      WorkerResetMessage(id: _requiredInt(message, 'id', 'reset'));
}

class WorkerResetResolverMessage extends WorkerMessage {
  WorkerResetResolverMessage({required int id}) : super(id: id);

  factory WorkerResetResolverMessage.fromJson(JsonMap message) =>
      WorkerResetResolverMessage(
        id: _requiredInt(message, 'id', 'reset_resolver'),
      );
}

class WorkerBuildMessage extends WorkerMessage {
  WorkerBuildMessage(this.request) : super(id: request.id);

  final WorkerBuildRequest request;
}

class WorkerBuildBatchMessage extends WorkerMessage {
  WorkerBuildBatchMessage({required int id, required this.requests})
    : super(id: id);

  factory WorkerBuildBatchMessage.fromJson(JsonMap message) {
    final rawRequests = message['requests'];
    if (rawRequests is! List) {
      throw const FormatException('build_batch requests must be a list');
    }
    return WorkerBuildBatchMessage(
      id: _requiredInt(message, 'id', 'build_batch'),
      requests: [
        for (final rawRequest in rawRequests)
          WorkerBuildRequest.fromJson(
            _jsonMap(rawRequest, 'build_batch request'),
          ),
      ],
    );
  }

  final List<WorkerBuildRequest> requests;
}

class UnsupportedWorkerMessage extends WorkerMessage {
  const UnsupportedWorkerMessage({required super.id, required this.type});

  final Object? type;
}

/// A validated build request, including nested optional-build requests.
class WorkerBuildRequest {
  WorkerBuildRequest({
    required this.id,
    required this.builder,
    required this.input,
    required this.kind,
    required this.allowedOutputs,
    required this.options,
    required this.phase,
    required this.instanceKey,
    required this.isRoot,
    required this.blockedAssets,
    required this.triggers,
  });

  factory WorkerBuildRequest.fromJson(JsonMap message) {
    final rawKind = message['kind'];
    final kind = rawKind == null
        ? null
        : _requiredStringValue(rawKind, 'build kind');
    if (kind != null && kind != 'normal' && kind != 'post_process') {
      throw FormatException('unsupported build kind: $kind');
    }
    final rawIsRoot = message['is_root'];
    final rawInstanceKey = message['instance_key'];
    final rawPhase = message['phase'];
    final rawOptions = message['options'];
    final options = rawOptions == null
        ? <String, dynamic>{}
        : _stringKeyedMap(rawOptions, 'build options');
    final rawAllowedOutputs = message['allowed_outputs'];
    final rawBlockedAssets = message['blocked_assets'];
    final rawTriggers = message['triggers'];
    return WorkerBuildRequest(
      id: _requiredInt(message, 'id', 'build'),
      builder: _requiredString(message, 'builder', 'build'),
      input: _requiredString(message, 'input', 'build'),
      kind: kind,
      allowedOutputs: _stringList(
        rawAllowedOutputs ?? const <dynamic>[],
        'build allowed_outputs',
      ),
      options: options,
      phase: rawPhase is num ? rawPhase.toInt() : 0,
      instanceKey: rawInstanceKey is String && rawInstanceKey.isNotEmpty
          ? rawInstanceKey
          : null,
      isRoot: rawIsRoot is bool ? rawIsRoot : true,
      blockedAssets: _stringList(
        rawBlockedAssets ?? const <dynamic>[],
        'build blocked_assets',
      ),
      triggers: _triggerList(
        rawTriggers ?? const <dynamic>[],
        'build triggers',
      ),
    );
  }

  final int id;
  final String builder;
  final String input;
  final String? kind;
  final List<String> allowedOutputs;
  final Map<String, dynamic> options;
  final int phase;
  final String? instanceKey;
  final bool isRoot;
  final List<String> blockedAssets;
  final List<WorkerBuildTrigger> triggers;

  bool get isPostProcess => kind == 'post_process';
}

class WorkerBuildTrigger {
  const WorkerBuildTrigger({required this.kind, required this.value});

  factory WorkerBuildTrigger.fromJson(Object? raw) {
    final message = _jsonMap(raw, 'build trigger');
    return WorkerBuildTrigger(
      kind: _requiredString(message, 'kind', 'build trigger'),
      value: _requiredString(message, 'value', 'build trigger'),
    );
  }

  final String kind;
  final String value;
}

int _requiredInt(JsonMap message, String key, String type) {
  final value = message[key];
  if (value is! int) {
    throw FormatException('$type requires an integer $key');
  }
  return value;
}

String _requiredString(JsonMap message, String key, String type) =>
    _requiredStringValue(message[key], '$type $key');

String _requiredStringValue(Object? value, String label) {
  if (value is! String) {
    throw FormatException('$label must be a string');
  }
  return value;
}

JsonMap _jsonMap(Object? value, String label) {
  if (value is! Map) {
    throw FormatException('$label must be an object');
  }
  final result = <String, dynamic>{};
  for (final entry in value.entries) {
    if (entry.key is! String) {
      throw FormatException('$label keys must be strings');
    }
    result[entry.key as String] = entry.value;
  }
  return result;
}

Map<String, dynamic> _stringKeyedMap(Object? value, String label) =>
    _jsonMap(value, label);

List<String> _stringList(Object? value, String label) {
  if (value is! List) {
    throw FormatException('$label must be a list');
  }
  return [for (final item in value) _requiredStringValue(item, '$label item')];
}

List<WorkerBuildTrigger> _triggerList(Object? value, String label) {
  if (value is! List) {
    throw FormatException('$label must be a list');
  }
  return [for (final item in value) WorkerBuildTrigger.fromJson(item)];
}

class FrameReader {
  FrameReader(Stream<List<int>> input) : _iterator = StreamIterator(input);

  final StreamIterator<List<int>> _iterator;
  final List<int> _buffer = <int>[];

  Future<JsonMap?> next() async {
    await _fill(4);
    if (_buffer.isEmpty) return null;

    final length = ByteData.sublistView(
      Uint8List.fromList(_buffer.sublist(0, 4)),
    ).getUint32(0, Endian.big);
    if (length > _maxFrameLength) {
      throw const FormatException('IPC frame exceeds the 256 MiB safety limit');
    }
    await _fill(4 + length);
    final payload = Uint8List.fromList(_buffer.sublist(4, 4 + length));
    _buffer.removeRange(0, 4 + length);
    if (_hasBinaryAssetResponseMagic(payload)) {
      return _decodeBinaryAssetResponse(payload);
    }
    final decoded = jsonDecode(utf8.decode(payload));
    if (decoded is! Map) {
      throw FormatException('IPC payload must be a JSON object');
    }
    return decoded.cast<String, dynamic>();
  }

  bool _hasBinaryAssetResponseMagic(Uint8List payload) {
    if (payload.length < _binaryAssetResponseMagic.length) return false;
    for (var index = 0; index < _binaryAssetResponseMagic.length; index++) {
      if (payload[index] != _binaryAssetResponseMagic[index]) return false;
    }
    return true;
  }

  JsonMap _decodeBinaryAssetResponse(Uint8List payload) {
    const envelopeHeaderLength = 8;
    if (payload.length < envelopeHeaderLength) {
      throw const FormatException('Binary IPC envelope is truncated');
    }
    final metadataLength = ByteData.sublistView(
      payload,
      _binaryAssetResponseMagic.length,
      envelopeHeaderLength,
    ).getUint32(0, Endian.big);
    final bytesStart = envelopeHeaderLength + metadataLength;
    if (bytesStart > payload.length) {
      throw const FormatException('Binary IPC metadata is truncated');
    }

    final decoded = jsonDecode(
      utf8.decode(payload.sublist(envelopeHeaderLength, bytesStart)),
    );
    if (decoded is! Map) {
      throw const FormatException('Binary IPC metadata must be a JSON object');
    }
    final message = <String, dynamic>{...decoded.cast<String, dynamic>()};
    if (message['type'] != 'asset_response' || message['encoding'] != 'raw') {
      throw const FormatException('Unsupported binary IPC envelope');
    }
    final expectedLength = message['length'];
    final actualLength = payload.length - bytesStart;
    if (expectedLength is! int || expectedLength != actualLength) {
      throw const FormatException('Binary IPC payload length mismatch');
    }
    // Keep the frame buffer alive through the view and defer the defensive
    // copy until a builder receives the bytes from the shared read cache.
    message['bytes'] = Uint8List.sublistView(payload, bytesStart);
    return message;
  }

  Future<void> _fill(int length) async {
    while (_buffer.length < length) {
      if (!await _iterator.moveNext()) {
        if (_buffer.isEmpty) return;
        throw const FormatException('Unexpected EOF in IPC frame');
      }
      _buffer.addAll(_iterator.current);
    }
  }
}

class FrameWriter {
  FrameWriter(this._output);

  final IOSink _output;

  Future<void> send(JsonMap message) async {
    final payload = utf8.encode(jsonEncode(message));
    if (payload.length > _maxFrameLength) {
      throw StateError('IPC frame exceeds the 256 MiB safety limit');
    }
    final header = ByteData(4)..setUint32(0, payload.length, Endian.big);
    _output.add(header.buffer.asUint8List());
    _output.add(payload);
    await _output.flush();
  }

  Future<void> sendBuildResult(JsonMap message) async {
    final outputBytes = <Uint8List>[];
    final metadata = _binaryBuildResultMetadata(message, outputBytes);
    await _sendBinary(_binaryBuildResultMagic, metadata, outputBytes);
  }

  JsonMap _binaryBuildResultMetadata(
    JsonMap message,
    List<Uint8List> outputBytes,
  ) {
    final metadata = <String, dynamic>{...message, 'encoding': 'raw'};
    switch (message['type']) {
      case 'build_result':
        metadata['outputs'] = _binaryOutputDescriptors(
          message['outputs'],
          outputBytes,
        );
      case 'build_batch_result':
        final rawResults = message['results'];
        if (rawResults is! List) {
          throw const FormatException(
            'build_batch_result results must be a list',
          );
        }
        metadata['results'] = <JsonMap>[
          for (final rawResult in rawResults)
            _binaryBuildResultMetadata(
              _asJsonMap(rawResult, 'build result'),
              outputBytes,
            ),
        ];
      default:
        throw FormatException(
          'Unsupported binary build result type: ${message['type']}',
        );
    }
    return metadata;
  }

  List<JsonMap> _binaryOutputDescriptors(
    dynamic rawOutputs,
    List<Uint8List> outputBytes,
  ) {
    if (rawOutputs is! List) {
      throw const FormatException('build_result outputs must be a list');
    }
    return <JsonMap>[
      for (final rawOutput in rawOutputs)
        _binaryOutputDescriptor(
          _asJsonMap(rawOutput, 'build output'),
          outputBytes,
        ),
    ];
  }

  JsonMap _binaryOutputDescriptor(JsonMap output, List<Uint8List> outputBytes) {
    final asset = output['asset'];
    final rawBytes = output['bytes'];
    if (asset is! String || rawBytes is! List) {
      throw const FormatException(
        'build output must contain an asset and bytes list',
      );
    }
    final bytes = rawBytes is Uint8List
        ? rawBytes
        : Uint8List.fromList(List<int>.from(rawBytes));
    outputBytes.add(bytes);
    return <String, dynamic>{'asset': asset, 'length': bytes.length};
  }

  JsonMap _asJsonMap(dynamic value, String label) {
    if (value is! Map) {
      throw FormatException('$label must be an object');
    }
    return value.cast<String, dynamic>();
  }

  Future<void> _sendBinary(
    List<int> magic,
    JsonMap metadata,
    List<Uint8List> rawChunks,
  ) async {
    final metadataPayload = Uint8List.fromList(
      utf8.encode(jsonEncode(metadata)),
    );
    var rawLength = 0;
    for (final chunk in rawChunks) {
      rawLength += chunk.length;
    }
    final payloadLength = magic.length + 4 + metadataPayload.length + rawLength;
    if (payloadLength > _maxFrameLength) {
      throw StateError('IPC frame exceeds the 256 MiB safety limit');
    }
    final header = ByteData(4)..setUint32(0, payloadLength, Endian.big);
    final metadataLength = ByteData(4)
      ..setUint32(0, metadataPayload.length, Endian.big);
    _output.add(header.buffer.asUint8List());
    _output.add(magic);
    _output.add(metadataLength.buffer.asUint8List());
    _output.add(metadataPayload);
    for (final chunk in rawChunks) {
      _output.add(chunk);
    }
    await _output.flush();
  }
}

class RpcSession {
  RpcSession(
    this.reader,
    this.writer, {
    required this.buildId,
    required this.phase,
    required this.postProcess,
    this.onControlMessage,
  });

  final FrameReader reader;
  final FrameWriter writer;
  final int buildId;
  final int phase;
  final bool postProcess;
  final RpcControlMessageHandler? onControlMessage;
  int _nextId = 1000;

  Future<JsonMap> call(String op, Map<String, dynamic> parameters) async {
    final id = _nextId++;
    await writer.send(<String, dynamic>{
      'v': 1,
      'type': 'asset_request',
      'id': id,
      'op': op,
      ...parameters,
      'build_id': buildId,
      // Rust applies the same phase-aware logical view as the Dart adapter.
      // Keeping these on every asset request also protects custom workers
      // which do not consume the build request's blocked_assets hint.
      'phase': phase,
      'kind': postProcess ? 'post_process' : 'normal',
    });
    while (true) {
      final response = await reader.next();
      if (response == null) {
        throw const FormatException('Rust frontend exited during RPC');
      }
      if (response['type'] == 'build') {
        final handler = onControlMessage;
        if (handler == null) {
          throw StateError('Unexpected nested build request during asset RPC');
        }
        await handler(WorkerBuildRequest.fromJson(response));
        continue;
      }
      if (response['type'] != 'asset_response' || response['id'] != id) {
        throw StateError(
          'Unexpected message while waiting for asset RPC: $response',
        );
      }
      if (response['ok'] != true) {
        throw StateError(response['error']?.toString() ?? 'Asset RPC failed');
      }
      return response;
    }
  }
}
